### PR TITLE
Alternate milkdrop preset format

### PR DESCRIPTION
**Pull request by mbellew**
An easier to parse file format for milkdrop presets.  Obviously, not backward compatible (but convertable).  My goal was to be more readable/editable but could be mechanically translated to the current format.   Below is an example of first proposal  (abridged for length).

```
preset 'AdamFx 2 Geiss -Somewhat Distort Me 3_1' {
  MILKDROP_PRESET_VERSION=201
  PSVERSION=2
  PSVERSION_WARP=2
  PSVERSION_COMP=2
  fRating=4.000000
  fGammaAdj=1.000
  fDecay=1.000
  fVideoEchoZoom=2.000
  fVideoEchoAlpha=0.000
  nVideoEchoOrientation=0
  nWaveMode=1
  bAdditiveWaves=0
  bWaveDots=0
  bWaveThick=0
  bModWaveAlphaByVolume=1
  bMaximizeWaveColor=1
  bTexWrap=1
  bDarkenCenter=0
  bRedBlueStereo=0
  bBrighten=0
  bDarken=0
  bSolarize=0
  bInvert=0
  fWaveAlpha=8.200
  fWaveScale=1.013
  fWaveSmoothing=0.900
  fWaveParam=-0.280
  fModWaveAlphaStart=1.050
  fModWaveAlphaEnd=1.650
  fWarpAnimSpeed=1.000
  fWarpScale=2.853
  fZoomExponent=3.60000
  fShader=0.000
  zoom=1.02500
  rot=-0.02000
  cx=0.500
  cy=0.500
  dx=0.00000
  dy=0.00000
  warp=0.30900
  sx=1.00000
  sy=1.00000
  wave_r=0.700
  wave_g=0.650
  wave_b=0.700
  wave_x=0.500
  wave_y=0.500
  ob_size=0.010
  ob_r=0.000
  ob_g=0.000
  ob_b=0.000
  ob_a=0.500
  ib_size=0.010
  ib_r=0.250
  ib_g=0.250
  ib_b=0.250
  ib_a=0.500
  nMotionVectorsX=31.200
  nMotionVectorsY=2.280
  mv_dx=0.000
  mv_dy=0.000
  mv_l=2.500
  mv_r=1.000
  mv_g=1.000
  mv_b=0.800
  mv_a=0.000
  b1n=0.000
  b2n=0.000
  b3n=0.000
  b1x=1.000
  b2x=1.000
  b3x=1.000
  b1ed=0.250

  per_frame = {
    wave_r =  wave_r + 0.200*( 0.60*sin(0.933*time) + 0.40*sin(1.045*time) );
    wave_g =  wave_g + 0.200*( 0.60*sin(0.900*time) + 0.40*sin(0.956*time) );
    wave_b =  wave_b + 0.200*( 0.60*sin(0.910*time) + 0.40*sin(0.920*time) );
    zoom =  zoom + 0.023*( 0.60*sin(0.339*time) + 0.40*sin(0.276*time) );
    rot =  rot + 0.030*( 0.60*sin(0.381*time) + 0.40*sin(0.579*time) );
  }

  per_pixel = {
    zoom = zoom+0.03*sin((x*2-1)*4+time*1.63)+0.03*sin((y*2-1)*3+time*1.37)-0.1*sin(rad*0.1+time*1.6);
  }

  comp_hlsl = """
    shader_body
    {
        ret = tex2D(sampler_main, uv).xyz;    
    }
"""

  warp_hlsl = """
    shader_body
    {
        // sample previous frame
        ret = tex2D( sampler_main, uv ).xyz;
    
        // take the difference between the crisp and blurred images,
        // then add it back into the image.  Creates spots and stripes over time.
        ret += (ret - GetBlur2(uv))*0.3;
        ret *= 0.9;
    
       // add noise:
       float2 dither_uv = uv_orig*texsize.xy*texsize_noise_lq.zw * 0.4 + rand_frame.xy;
       ret += (tex2D(sampler_noise_lq, dither_uv).xyz-0.5)/256.0 
                   * 122 * saturate((treb_att-1)+0.4);
    
       // desaturate over time, to keep the globs white
       ret = lerp(ret, lum(ret), 0.2);
    }
"""
}

shape '0' {
  enabled=0
  sides=4
  additive=0
  thickOutline=0
  textured=0
  num_inst=1
  x=0.500
  y=0.500
  rad=0.10000
  ang=0.00000
  tex_ang=0.00000
  tex_zoom=1.00000
  r=1.000
  g=0.000
  b=0.000
  a=1.000
  r2=0.000
  g2=1.000
  b2=0.000
  a2=0.000
  border_r=1.000
  border_g=1.000
  border_b=1.000
  border_a=0.100
}
// ... SHAPES DELETED

wave '0' {
  enabled=0
  samples=512
  sep=0
  bSpectrum=0
  bUseDots=0
  bDrawThick=0
  bAdditive=0
  scaling=1.00000
  smoothing=0.50000
  r=1.000
  g=1.000
  b=1.000
  a=1.000
}
// ... WAVES DELETED
```

### CODE REVIEW

**Comment by revmischa** [comment]
Regarding the sample – is there a reason some values are `1` and some are `1.00000`? does it make a difference? also weird that the shape and wave ids are quoted
cool though! much nicer. wonder if r/g/b values could be expressed as a sort of tuple?

**Comment by mbellew** [comment]
No reason for the number formatting, that's just copied over from the source .milk file.  I love the idea of supporting rgb tuples and/or css style.

Now onto looking at the continuous integration checks...

File: src/libprojectM/MilkdropPresetFactory/SaltWaterTaffyPreset.cpp [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+        else if (value == "\"\"\"")
+        {
+            std::string shader="comp";
+            std::string lang="hlsl";
+            if (name != "comp_hlsl" && name != "comp_glsl_330" && name != "warp_hlsl" && name != "warp_glsl_330")
+            {
+                if (PARSE_DEBUG) printf("parse taffy: shader name should be one of comp_hlsl, comp_glsl_330, warp, or warp_glsl_330: %s\n", line.c_str());
+                return PROJECTM_PARSE_ERROR;
+            }
+            if (name == "comp_glsl_330" || name == "warp_glsl_330")
```
Comment (revmischa) [inline src/libprojectM/MilkdropPresetFactory/SaltWaterTaffyPreset.cpp]: the versions are highly likely to change. embedding these constants deep down in here is a very bad idea.

File: src/libprojectM/MilkdropPresetFactory/SaltWaterTaffyPreset.cpp [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+                lang = "glsl_330";
+            if (starts_with(name, "warp"))
+                shader = "warp";
+            // TODO enable this with the glsl branch
+            // os << shader << "_lang=" << lang << name << std::endl;
```
Comment (revmischa) [inline src/libprojectM/MilkdropPresetFactory/SaltWaterTaffyPreset.cpp]: glsl branch is all merged

File: taffy.py [hunk 3]
```diff
@@ -0,0 +1,445 @@
+#! /usr/bin/python3
```
Comment (revmischa) [inline taffy.py]: no space after #!

File: taffy.py [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+    def get_wave(self, i):
+        i = int(i)
+        while len(self.waves) <= int(i):
+            self.waves.append(Shape(len(self.waves)))
+        return self.waves[i]
+
+
+# 'parse' right hand side of expression, allow substition of identifiers
+def scan_expr(expr, ident_callback):
```
Comment (revmischa) [inline taffy.py]: this is pretty wild.. you should look at https://kite.com/python/docs/pyparsing

Comment (mbellew) [reply taffy.py]: Thanks for feedback.  The identifier substitution was kind of a experiment, I might just remove it.  Thanks for the pyparsing suggestion.

**Comment by mbellew** [comment]
May look at building a cleaner parser

**Pull request closed without merging**